app/src/main/kotlin/GridWorld/GridworldMDP.kt

```diff
@@ -29,47 +32,48 @@ class GridworldMDP(
     }
 
     override fun reward(previousState: GridworldState?, action: GridworldAction?, state: GridworldState): Double {
-        var reward = 0.0
         for (r in rewards) {
             if (r.equals(state)) {
-                reward += r.value
+                return r.value
             }
         }
 
-        return reward
+        return 0.0
     }
 
     override fun transition(state: GridworldState, action: GridworldAction): GridworldState {
         if (state.isTerminal) {
             return state
         }
         else if (rewards.any { r -> r.equals(state)}) {
-            return GridworldState(state.x, state.y, true)
+            return _rewardStates.getOrDefault(state, GridworldState(state.x, state.y, true))
         }
 
         // if target is out of bounds, return current state
-        val targetNeighbour = state.ResolveNeighbour(action, xSize, ySize) ?:
+        val targetNeighbour = state.resolveNeighbour(action, xSize, ySize) ?:
             return state
 
-        val nonTargetNeighbours = mutableListOf<GridworldState>()
-
-        for (a in GridworldAction.values().toList().minus(action)) {
-            val possibleNeighbour = state.ResolveNeighbour(a, xSize, ySize)
-            if (possibleNeighbour != null) {
-                nonTargetNeighbours.add(possibleNeighbour)
-            }
-        }
-
         if (Math.random() < transitionProbability){
             return targetNeighbour
         } else {
-            return nonTargetNeighbours.random()
-        }
+            var nonTargetNeighbours : MutableList<GridworldState>? = null
```

Comment:
I was under the impression `nonTargetNeighbours.random()` just returns a random action in the possible actions that the agent will go to. Seems like the update is now returning a mutable list... can you explain the rationale here?

Reply:
No, it's still only returning a single action on line 71. This refactoring moves a bunch of code such as creating a mutable list object and resolving possible neighbours into the condition where the target neighbour is not chosen. Originally, we spend a lot of time doing all the aforementioned work which is unnecessary if we ultimately end up choosing the target neighbour anyway.